<a href="https://colab.research.google.com/github/rejina0914/Worksheet/blob/main/Week_5_Workshop(AI).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

• To - Do - 1:
1. Read and Observe the Dataset.
2. Print top(5) and bottom(5) of the dataset {Hint: pd.head and pd.tail}.
3. Print the Information of Datasets. {Hint: pd.info}.
4. Gather the Descriptive info about the Dataset. {Hint: pd.describe}
5. Split your data into Feature (X) and Label (Y).

In [2]:
import pandas as pd
df = pd.read_csv('/content/drive/MyDrive/Worksheet-2(AI)/student.csv')
print("Top 5 rows:")
print(df.head())

print("\nBottom 5 rows:")
print(df.tail())

print("\nDataset Information:")
df.info()

print("\nDescriptive Info:")
print(df.describe())

# Step 2: Split the data into features (X) and Label (Y)
X = df[['Math', 'Reading']]
Y = df['Writing']

print("\nFeatures (X):")
print(X.head())

print("\nLabel (Y):")
print(Y.head())

Top 5 rows:
   Math  Reading  Writing
0    48       68       63
1    62       81       72
2    79       80       78
3    76       83       79
4    59       64       62

Bottom 5 rows:
     Math  Reading  Writing
995    72       74       70
996    73       86       90
997    89       87       94
998    83       82       78
999    66       66       72

Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   Math     1000 non-null   int64
 1   Reading  1000 non-null   int64
 2   Writing  1000 non-null   int64
dtypes: int64(3)
memory usage: 23.6 KB

Descriptive Info:
              Math      Reading      Writing
count  1000.000000  1000.000000  1000.000000
mean     67.290000    69.872000    68.616000
std      15.085008    14.657027    15.241287
min      13.000000    19.000000    14.000000
25%      58.000000    60.750000    58.000000
50%      68.0

To - Do - 2:
1. To make the task easier - let’s assume there is no bias or intercept.
2. Create the following matrices:

Y = WTX

In [5]:
import  numpy as np
X = df[['Math', 'Reading']].values.T

Y = df['Writing'].values

d = X.shape[0]
W = np.zeros(d)

Y_pred = np.dot(W.T, X)

print("Shape of X (features x samples):", X.shape)
print("Shape of W (weights):", W.shape)
print("Shape of Y_pred (predictions):", Y_pred.shape)

print("First 5 predicted values:", Y_pred[:5])

Shape of X (features x samples): (2, 1000)
Shape of W (weights): (2,)
Shape of Y_pred (predictions): (1000,)
First 5 predicted values: [0. 0. 0. 0. 0.]


To - Do - 3:
1. Split the dataset into training and test sets.
2. You can use an 80-20 or 70-30 split, with 80% (or 70%) of the data used for training and the rest
for testing.

In [6]:
from sklearn.model_selection import train_test_split

X = df[['Math', 'Reading']]
Y = df['Writing']

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of Y_train:", Y_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of Y_test:", Y_test.shape)

Shape of X_train: (800, 2)
Shape of Y_train: (800,)
Shape of X_test: (200, 2)
Shape of Y_test: (200,)


To - Do - 4:
Feel free to build your own code or complete the following code:
Building a Cost Function:

In [15]:
#Define the cost function
import numpy as np
def cost_function(X, Y, W):
  """ Parameters:
This function finds the Mean Square Error.
Input parameters:
X: Feature Matrix
Y: Target Matrix
W: Weight Matrix
Output Parameters:
cost: accumulated mean square error.
"""
  m = len(Y)
  J = np.sum((X.dot(W)-Y) ** 2)/(2 * m)
  return J


To - Do - 5: Make sure your code at To - Do - 4 passed the following test case: Testing a Cost Function:

In [19]:
import numpy as np
# Test case
X_test = np.array([[1, 2], [3, 4], [5, 6]])
Y_test = np.array([3, 7, 11])
W_test = np.array([1, 1])
cost = cost_function(X_test, Y_test, W_test)
if cost == 0:
  print("Proceed Further")
else:
  print("something went wrong: Reimplement a cost function")
print("Cost function output:", cost_function(X_test, Y_test, W_test))

Proceed Further
Cost function output: 0.0


To - Do - 6:
Implement your code for Gradient Descent; Either fill the following code or write your own:

Gradient Descent from Scratch:


In [20]:
from re import X
def gradient_descent(X, Y, W, alpha, iterations):
  """
Perform gradient descent to optimize the parameters of a linear regression model.
Parameters:
X (numpy.ndarray): Feature matrix (m x n).
Y (numpy.ndarray): Target vector (m x 1).
W (numpy.ndarray): Initial guess for parameters (n x 1).
alpha (float): Learning rate.
iterations (int): Number of iterations for gradient descent.
Returns:
tuple: A tuple containing the final optimized parameters (W_update) and the history of cost values
.
W_update (numpy.ndarray): Updated parameters (n x 1).
cost_history (list): History of cost values over iterations.
"""
  # Initialize cost history
  cost_history = [0] * iterations
  # Number of samples
  m = len(Y)
  for iteration in range(iterations):
    # Step 1: Hypothesis Values
    Y_pred = X.dot(W)
    # Step 2: Difference between Hypothesis and Actual Y
    loss = Y_pred - Y
    # Step 3: Gradient Calculation
    dw = (X.T.dot(loss)) / (m)
    # Step 4: Updating Values of W using Gradient
    W_update = W - alpha * dw
    # Step 5: New Cost Value
    cost = cost_function(X, Y, W_update)
    cost_history[iteration] = cost
  return W_update, cost_history

To - Do - 7:
Make sure following Test Case is passe by your code from To - Do - 6 or your Gradient Descent
Implementation:

In [21]:
# Generate random test data
np.random.seed(0) # For reproducibility
X = np.random.rand(100, 3) # 100 samples, 3 features
Y = np.random.rand(100)
W = np.random.rand(3) # Initial guess for parameters
# Set hyperparameters
alpha = 0.01
iterations = 1000
# Test the gradient_descent function
final_params, cost_history = gradient_descent(X, Y, W, alpha, iterations)
# Print the final parameters and cost history
print("Final Parameters:", final_params)
print("Cost History:", cost_history)

Final Parameters: [0.3996496  0.92745322 0.09826523]
Cost History: [np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float64(0.10711197094660153), np.float

To - Do - 8:
Implementation of RMSE in the Code - Complete the following code or write your own:

Code for RMSE:

In [23]:
# Model Evaluation - RMSE
def rmse(Y, Y_pred):
  """
This Function calculates the Root Mean Squres.
Input Arguments:
Y: Array of actual(Target) Dependent Varaibles.
Y_pred: Array of predeicted Dependent Varaibles.
Output Arguments:
rmse: Root Mean Square.
"""
  rmse = np.sqrt(np.mean((Y - Y_pred) ** 2))
  return rmse


To - Do - 9 - Implementation in the Code:
Complete the following code or write your own for r2 loss:
Code for R-Squared Error:

In [24]:
# Model Evaluation - R2
def r2(Y, Y_pred):
  """
This Function calculates the R Squared Error.
Input Arguments:
Y: Array of actual(Target) Dependent Varaibles.
Y_pred: Array of predeicted Dependent Varaibles.
Output Arguments:
rsquared: R Squared Error.
"""
  mean_y = np.mean(Y)
  ss_tot = np.sum((Y - mean_y) ** 2)
  ss_res = np.sum((Y - Y_pred) ** 2)
  r2 = 1 - (ss_res / ss_tot)
  return r2

• To - Do - 10:
We will define a function that:
1. Loads the data and splits it into training and test sets.
2. Prepares the feature matrix (X) and target vector (Y).
3. Defines the weight matrix (W) and initializes the learning rate and number of iterations.
4. Calls the gradient descent function to learn the parameters.
5. Evaluates the model using RMSE and R2
.

In [25]:
# Main Function
def main():
# Step 1: Load the dataset
 df = pd.read_csv("/content/drive/MyDrive/Worksheet-2(AI)/student.csv")

# Step 2: Split the data into features (X) and target (Y)
 X = df[["Math", "Reading"]].values # Features: Math and Reading marks
 Y = df["Writing"].values # Target: Writing marks

# Step 3: Split the data into training and test sets (80% train, 20% test)
 X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

# Step 4: Initialize weights (W) to zeros, learning rate and number of iterations
 W = np.zeros(X_train.shape[1]) # Initialize weights
 alpha = 0.00001 # Learning rate
 iterations = 1000 # Number of iterations for gradient descent

# Step 5: Perform Gradient Descent
 W_optimal, cost_history = gradient_descent(X_train, Y_train, W, alpha, iterations)

# Step 6: Make predictions on the test set
 Y_pred = np.dot(X_test, W_optimal)

# Step 7: Evaluate the model using RMSE and R-Squared
 model_rmse = rmse(Y_test, Y_pred)
 model_r2 = r2(Y_test, Y_pred)

# Step 8: Output the results
 print("Final Weights:", W_optimal)
 print("Cost History (First 10 iterations):", cost_history[:10])
 print("RMSE on Test Set:", model_rmse)
 print("R-Squared on Test Set:", model_r2)
# Execute the main function
if __name__ == "__main__":
  main()

Final Weights: [0.04797833 0.05020199]
Cost History (First 10 iterations): [np.float64(2013.165570783755), np.float64(2013.165570783755), np.float64(2013.165570783755), np.float64(2013.165570783755), np.float64(2013.165570783755), np.float64(2013.165570783755), np.float64(2013.165570783755), np.float64(2013.165570783755), np.float64(2013.165570783755), np.float64(2013.165570783755)]
RMSE on Test Set: 63.36563528655014
R-Squared on Test Set: -15.04041794561271


To - Do - 11 - Present your finding:
1. Did your Model Overfitt, Underfitts, or performance is acceptable.

 Ans: When alpha = 0.00001, The model is underfitting as the RMSE value is very high and R-square is negative. The cost history is slao constant which states Gradient descent did not converge properly.


2. Experiment with different value of learning rate, making it higher and lower, observe the result.

alpha = 0.0001: Final Weights: [0.47978325 0.50201988] Cost History (First 10 iterations): [np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211), np.float64(17.81379717752211)] RMSE on Test Set: 6.092665443799174 R-Squared on Test Set: 0.851706281452247

When alpha value is increased from 0.00001 to 0.0001, the model perfomance improved. The RMSE reduced from 63.36 to 6.09, and the R² score increased from −15.04 to 0.85, indicating that the model learned effectively. So, the model perfomance is acceptable neither underfitting nor overfitting.

aplha = 0.001: Final Weights: [4.7978325 5.02019875] Cost History (First 10 iterations): [np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246), np.float64(191077.53315577246)] RMSE on Test Set: 615.8660062668006 R-Squared on Test Set: -1514.236975362205

The model is unstable and performs poorly.